File: python/src/linksql/query.py
"""High-level query executor: the public face of the engine.

A :class:`Database` bundles a links store with a name registry and exposes a
single :meth:`Database.query` method that accepts LiNo text. The text is parsed
into a restriction and a substitution, the operation runs, and a structured
report comes back describing what matched and what changed.

This module is a faithful, behaviourally-identical port of ``src/query.js`` from
the JavaScript reference implementation.
"""

from __future__ import annotations

import dataclasses
from dataclasses import dataclass, field
from typing import Callable

from .lino import Link as LinkNode
from .lino import LinoSyntaxError, Node, Ref, parse, serialize
from .names import Names
from .store import Link, LinksStore
from .substitution import (
    Context,
    RawResult,
    Row,
    SubstitutionError,
    execute,
    match_restriction,
)


class QueryError(Exception):
    """Error raised when query text is well-formed LiNo but not a valid query."""


@dataclass
class SplitQuery:
    """A parsed query split into a restriction and an optional substitution."""

    restriction: list[Node]
    substitution: list[Node] | None


@dataclass
class MatchRow:
    """One row of a query's matches: the matched links and the bindings."""

    links: list[Link]
    binding: dict[str, int]


@dataclass
class QueryReport:
    """A structured report describing what a query matched and changed.

    Every :class:`~linksql.store.Link` in the report is a copy, so mutating the
    store after a query never aliases the reported links.
    """

    operation: str
    matched: list[MatchRow] = field(default_factory=list)
    created: list[Link] = field(default_factory=list)
    updated: list[Link] = field(default_factory=list)
    deleted: list[Link] = field(default_factory=list)


@dataclass
class Introspection:
    """A snapshot of the database for introspection tooling.

    ``link_count`` is the Python-idiomatic name for the JavaScript
    ``linkCount``; ``names`` is a list of ``{"name", "index"}`` dicts and
    ``links`` enumerates every stored link.
    """

    link_count: int
    names: list[dict[str, object]]
    links: list[Link]


def patterns_of(node: Node) -> list[Node]:
    """Extract the pattern list from a restriction/substitution wrapper node."""
    if not isinstance(node, LinkNode) or node.id is not None:
        message = (
            "A restriction or substitution must be a parenthesised list of patterns"
        )
        raise QueryError(message)
    return node.values


def split_query(nodes: list[Node]) -> SplitQuery:
    """Split parsed LiNo nodes into a restriction and optional substitution.

    One top-level value is a read (restriction only). Two values are the
    canonical ``(restriction) (substitution)`` form. Anything else is ambiguous.
    """
    if len(nodes) == 0:
        return SplitQuery(restriction=[], substitution=None)
    if len(nodes) == 1:
        return SplitQuery(restriction=patterns_of(nodes[0]), substitution=None)
    if len(nodes) == 2:
        return SplitQuery(
            restriction=patterns_of(nodes[0]),
            substitution=patterns_of(nodes[1]),
        )
    message = 'A query must be "(restriction)" or "(restriction) (substitution)"'
    raise QueryError(message)


def _link_to_node(link: Link) -> LinkNode:
    """Convert a stored link to its canonical LiNo node."""
    return LinkNode(
        id=Ref(kind="number", value=link.index),
        values=[
            Ref(kind="number", value=link.source),
            Ref(kind="number", value=link.target),
        ],
    )


def link_to_lino(link: Link) -> str:
    """Serialise a link to canonical LiNo text, e.g. ``(3: 1 2)``."""
    return serialize(_link_to_node(link))


def _classify(raw: RawResult, read_only: bool) -> str:
    """Classify a change report into a human-friendly operation name."""
    if read_only:
        return "read"
    kinds = [
        name
        for name, links in (
            ("create", raw.created),
            ("update", raw.updated),
            ("delete", raw.deleted),
        )
        if links
    ]
    if len(kinds) == 0:
        return "noop"
    return kinds[0] if len(kinds) == 1 else "mixed"


def _copy_link(link: Link) -> Link:
    """Return an independent copy of a link so reports never alias the store."""
    return dataclasses.replace(link)


def _copy_row(row: Row) -> MatchRow:
    """Copy a join row into a report row, copying every matched link."""
    return MatchRow(
        links=[_copy_link(link) for link in row.links],
        binding=dict(row.binding),
    )


ChangeListener = Callable[[QueryReport], None]


class Database:
    """An associative database queried with the single substitution operation."""

    def __init__(self, *, auto_create: bool = True) -> None:
        self.store = LinksStore()
        self.names = Names(self.store, auto_create=auto_create)
        self.listeners: list[ChangeListener] = []

    @property
    def context(self) -> Context:
        """The execution context passed to the engine."""
        return Context(store=self.store, names=self.names)

    def on_change(self, listener: ChangeListener) -> Callable[[], None]:
        """Register a change listener; return an unsubscribe function."""
        self.listeners.append(listener)

        def unsubscribe() -> None:
            self.listeners = [
                existing for existing in self.listeners if existing is not listener
            ]

        return unsubscribe

    def emit(self, report: QueryReport) -> None:
        """Notify listeners about the changes produced by an operation."""
        if report.created or report.updated or report.deleted:
            for listener in self.listeners:
                listener(report)

    def query(self, text: str) -> QueryReport:
        """Run a LinksQL query expressed as LiNo text."""
        try:
            nodes = parse(text)
        except LinoSyntaxError as error:
            message = f"Invalid LiNo: {error}"
            raise QueryError(message) from error
        split = split_query(nodes)
        read_only = split.substitution is None
        try:
            if read_only:
                raw = RawResult(
                    matches=match_restriction(split.restriction, self.context)
                )
            else:
                assert split.substitution is not None
                raw = execute(split.restriction, split.substitution, self.context)
        except SubstitutionError as error:
            raise QueryError(str(error)) from error
        report = QueryReport(
            operation=_classify(raw, read_only),
            matched=[_copy_row(row) for row in raw.matches],
            created=[_copy_link(link) for link in raw.created],
            updated=[_copy_link(link) for link in raw.updated],
            deleted=[_copy_link(link) for link in raw.deleted],
        )
        self.emit(report)
        return report

    def links(self) -> list[Link]:
        """All stored links."""
        return self.store.all()

    def count(self) -> int:
        """Number of links currently stored."""
        return self.store.size

    def to_lino(self) -> str:
        """Serialise the whole database to canonical LiNo, one link per line."""
        return "\n".join(link_to_lino(link) for link in self.store.all())

    def import_lino(self, text: str) -> int:
        """Bulk-import links from LiNo text (each top-level link is created)."""
        nodes = parse(text)
        count = 0
        for node in nodes:
            report = self.query(f"() ({serialize(node)})")
            count += len(report.created)
        return count

    def introspect(self) -> Introspection:
        """Describe the database for introspection tooling."""
        return Introspection(
            link_count=self.store.size,
            names=[
                {"name": name, "index": index} for name, index in self.names.entries()
            ],
            links=self.store.all(),
        )

    def clear(self) -> None:
        """Remove every link and name."""
        self.store.clear()
        self.names.by_name.clear()
        self.names.by_index.clear()

File: csharp/src/LinksQL/Query.cs
using System.Globalization;

namespace LinksQL;

/// <summary>
/// The human-friendly classification of a query's effect.
/// </summary>
public enum Operation
{
    /// <summary>A read-only query (restriction only).</summary>
    Read,

    /// <summary>An operation that only created links.</summary>
    Create,

    /// <summary>An operation that only updated links.</summary>
    Update,

    /// <summary>An operation that only deleted links.</summary>
    Delete,

    /// <summary>An operation that created/updated/deleted in more than one way.</summary>
    Mixed,

    /// <summary>An operation that changed nothing.</summary>
    Noop,
}

/// <summary>
/// One join row of a query report: the concrete links matched and the variable
/// bindings shared across them.
/// </summary>
/// <param name="Links">The concrete links matched, in pattern order.</param>
/// <param name="Binding">The variable bindings as <c>name -&gt; index</c>.</param>
public sealed record MatchRow(IReadOnlyList<Link> Links, IReadOnlyDictionary<string, long> Binding);

/// <summary>
/// A structured, immutable report of an executed query.
/// </summary>
/// <param name="Operation">The classified operation.</param>
/// <param name="Matched">The join rows that matched the restriction.</param>
/// <param name="Created">The links created by this operation.</param>
/// <param name="Updated">The links updated by this operation (final structure).</param>
/// <param name="Deleted">The links removed by this operation.</param>
public sealed record QueryReport(
    Operation Operation,
    IReadOnlyList<MatchRow> Matched,
    IReadOnlyList<Link> Created,
    IReadOnlyList<Link> Updated,
    IReadOnlyList<Link> Deleted);

/// <summary>
/// A single named reference, as surfaced by <see cref="Database.Introspect"/>.
/// </summary>
/// <param name="Name">The human-readable label.</param>
/// <param name="Index">The link index it is bound to.</param>
public sealed record NamedReference(string Name, long Index);

/// <summary>
/// The introspection snapshot of a database: its link count, named references,
/// and links.
/// </summary>
/// <param name="LinkCount">The number of stored links.</param>
/// <param name="Names">Every <c>(name, index)</c> association in the registry.</param>
/// <param name="Links">Every stored link.</param>
public sealed record IntrospectionReport(
    long LinkCount,
    IReadOnlyList<NamedReference> Names,
    IReadOnlyList<Link> Links);

/// <summary>
/// High-level query executor helpers and the public <see cref="Database"/>.
/// </summary>
/// <remarks>
/// A <see cref="Database"/> bundles a links store with a name registry and
/// exposes a single <see cref="Database.Query(string)"/> method that accepts
/// LiNo text. The text is parsed into a restriction and a substitution, the
/// operation runs, and a structured report describing what matched and what
/// changed comes back.
/// </remarks>
public static class Query
{
    /// <summary>
    /// Extract the pattern list from a restriction/substitution wrapper node.
    /// </summary>
    /// <param name="node">A top-level <see cref="LinkNode"/>.</param>
    /// <returns>The contained pattern nodes.</returns>
    /// <exception cref="QueryError">
    /// Thrown when the node is not a parenthesised list of patterns (a link with
    /// no identity).
    /// </exception>
    public static IReadOnlyList<Node> PatternsOf(Node node)
    {
        if (node is not LinkNode link || link.Id is not null)
        {
            throw new QueryError(
                "A restriction or substitution must be a parenthesised list of patterns");
        }

        return link.Values;
    }

    /// <summary>
    /// Split parsed LiNo nodes into a restriction and an (optional) substitution.
    /// </summary>
    /// <param name="nodes">Parsed top-level nodes.</param>
    /// <returns>
    /// The restriction patterns and the substitution patterns, or
    /// <see langword="null"/> substitution for a read.
    /// </returns>
    /// <exception cref="QueryError">
    /// Thrown when there are more than two top-level nodes.
    /// </exception>
    public static (IReadOnlyList<Node> Restriction, IReadOnlyList<Node>? Substitution) SplitQuery(
        IReadOnlyList<Node> nodes)
    {
        ArgumentNullException.ThrowIfNull(nodes);

        if (nodes.Count == 0)
        {
            return (Array.Empty<Node>(), null);
        }

        if (nodes.Count == 1)
        {
            return (PatternsOf(nodes[0]), null);
        }

        if (nodes.Count == 2)
        {
            return (PatternsOf(nodes[0]), PatternsOf(nodes[1]));
        }

        throw new QueryError(
            "A query must be \"(restriction)\" or \"(restriction) (substitution)\"");
    }

    /// <summary>
    /// Convert a stored link to its canonical LiNo node.
    /// </summary>
    /// <param name="link">A link.</param>
    /// <returns>A <see cref="LinkNode"/> of the form <c>(index: source target)</c>.</returns>
    public static LinkNode LinkToNode(Link link)
    {
        ArgumentNullException.ThrowIfNull(link);

        return new LinkNode(
            new NumberRef(link.Index),
            new Node[] { new NumberRef(link.Source), new NumberRef(link.Target) });
    }

    /// <summary>
    /// Serialise a link to canonical LiNo text, e.g. <c>(3: 1 2)</c>.
    /// </summary>
    /// <param name="link">A link.</param>
    /// <returns>LiNo text.</returns>
    public static string LinkToLino(Link link) => Lino.Serialize(LinkToNode(link));

    /// <summary>
    /// Classify a change report into a human-friendly operation name.
    /// </summary>
    /// <param name="result">The execution result.</param>
    /// <param name="readOnly">Whether the query was read-only.</param>
    /// <returns>The classified <see cref="Operation"/>.</returns>
    public static Operation Classify(ExecutionResult result, bool readOnly)
    {
        ArgumentNullException.ThrowIfNull(result);

        if (readOnly)
        {
            return Operation.Read;
        }

        var kinds = new List<Operation>(3);
        if (result.Created.Count > 0)
        {
            kinds.Add(Operation.Create);
        }

        if (result.Updated.Count > 0)
        {
            kinds.Add(Operation.Update);
        }

        if (result.Deleted.Count > 0)
        {
            kinds.Add(Operation.Delete);
        }

        if (kinds.Count == 0)
        {
            return Operation.Noop;
        }

        return kinds.Count == 1 ? kinds[0] : Operation.Mixed;
    }
}

/// <summary>
/// An associative database queried with the single substitution operation.
/// </summary>
public sealed class Database
{
    private readonly List<Action<QueryReport>> _listeners = new();

    /// <summary>Initializes a new instance of the <see cref="Database"/> class.</summary>
    /// <param name="autoCreate">Auto-create missing named references.</param>
    public Database(bool autoCreate = true)
    {
        Store = new LinksStore();
        Names = new Names(Store, autoCreate);
    }

    /// <summary>Gets the backing links store.</summary>
    public LinksStore Store { get; }

    /// <summary>Gets the name registry.</summary>
    public Names Names { get; }

    /// <summary>Gets the execution context passed to the engine.</summary>
    private SubstitutionContext Context => new(Store, Names);

    /// <summary>
    /// Register a change listener (used by the subscription layer).
    /// </summary>
    /// <param name="listener">Callback for each change.</param>
    /// <returns>An unsubscribe action.</returns>
    public Action OnChange(Action<QueryReport> listener)
    {
        ArgumentNullException.ThrowIfNull(listener);

        _listeners.Add(listener);
        return () => _listeners.Remove(listener);
    }

    /// <summary>
    /// Notify listeners about the changes produced by an operation.
    /// </summary>
    /// <param name="report">The structured query report.</param>
    public void Emit(QueryReport report)
    {
        ArgumentNullException.ThrowIfNull(report);

        if (report.Created.Count > 0 || report.Updated.Count > 0 || report.Deleted.Count > 0)
        {
            foreach (var listener in _listeners.ToArray())
            {
                listener(report);
            }
        }
    }

    /// <summary>
    /// Run a LinksQL query expressed as LiNo text.
    /// </summary>
    /// <param name="text">The query, e.g. <c>() ((1 1))</c>.</param>
    /// <returns>A structured query report.</returns>
    /// <exception cref="QueryError">
    /// Thrown when the text is not valid LiNo, not a valid query shape, or the
    /// substitution cannot be carried out.
    /// </exception>
    public QueryReport Query(string text)
    {
        IReadOnlyList<Node> nodes;
        try
        {
            nodes = Lino.Parse(text);
        }
        catch (LinoSyntaxError error)
        {
            throw new QueryError($"Invalid LiNo: {error.Message}", error);
        }

        var (restriction, substitution) = LinksQL.Query.SplitQuery(nodes);
        var readOnly = substitution is null;
        ExecutionResult raw;
        try
        {
            if (readOnly)
            {
                raw = new ExecutionResult();
                raw.MatchesList.AddRange(Substitution.Match(restriction, Context));
            }
            else
            {
                raw = Substitution.Execute(restriction, substitution!, Context);
            }
        }
        catch (SubstitutionError error)
        {
            throw new QueryError(error.Message, error);
        }

        var report = new QueryReport(
            LinksQL.Query.Classify(raw, readOnly),
            raw.Matches.Select(ToMatchRow).ToArray(),
            raw.Created.ToArray(),
            raw.Updated.ToArray(),
            raw.Deleted.ToArray());
        Emit(report);
        return report;
    }

    /// <summary>All stored links.</summary>
    /// <returns>Every stored link, in insertion order.</returns>
    public IReadOnlyList<Link> Links() => Store.All();

    /// <summary>Number of links currently stored.</summary>
    /// <returns>The link count.</returns>
    public int Count() => Store.Size;

    /// <summary>
    /// Serialise the whole database to canonical LiNo, one link per line.
    /// </summary>
    /// <returns>LiNo text.</returns>
    public string ToLino() =>
        string.Join("\n", Store.All().Select(LinksQL.Query.LinkToLino));

    /// <summary>
    /// Bulk-import links from LiNo text (each top-level link is created).
    /// </summary>
    /// <param name="text">LiNo text of <c>(index: source target)</c> links.</param>
    /// <returns>The number of links imported.</returns>
    public int ImportLino(string text)
    {
        var nodes = Lino.Parse(text);
        var count = 0;
        foreach (var node in nodes)
        {
            var report = Query(
                string.Create(CultureInfo.InvariantCulture, $"() ({Lino.Serialize(node)})"));
            count += report.Created.Count;
        }

        return count;
    }

    /// <summary>
    /// Describe the database for introspection tooling.
    /// </summary>
    /// <returns>Link count, named references and the links themselves.</returns>
    public IntrospectionReport Introspect() =>
        new(
            Store.Size,
            Names.Entries().Select(entry => new NamedReference(entry.Name, entry.Index)).ToArray(),
            Store.All());

    /// <summary>Remove every link and name.</summary>
    public void Clear()
    {
        Store.Clear();
        Names.Clear();
    }

    /// <summary>Project an engine <see cref="Row"/> into an immutable report row.</summary>
    private static MatchRow ToMatchRow(Row row) =>
        new(
            row.Links.ToArray(),
            new Dictionary<string, long>(row.Binding, StringComparer.Ordinal));
}
